Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm

```diff
@@ -1874,6 +1878,93 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
     }
 }
 
+// MARK: webRequest
+
+bool WebExtensionContext::hasPermissionToSendWebRequestEvent(WebExtensionTab *tab, const URL& resourceURL, const ResourceLoadInfo& loadInfo)
+{
+    if (!tab)
+        return false;
+
+    if (!hasPermission(_WKWebExtensionPermissionWebRequest, tab))
+        return false;
+
+    auto tabURL = tab->url();
+    if (tabURL.isValid() && !hasPermission(tabURL, tab))
+        return false;
+
+    if (resourceURL.isValid() && !hasPermission(resourceURL, tab))
+        return false;
+
+    const URL& resourceLoadURL = loadInfo.originalURL;
+    if (resourceLoadURL.isValid() && !hasPermission(resourceLoadURL, tab))
+        return false;
+
+    return true;
+}
+
+void WebExtensionContext::resourceLoadDidSendRequest(WebPageProxyIdentifier pageID, const ResourceLoadInfo& loadInfo, const WebCore::ResourceRequest& request)
+{
+    auto tab = getTab(pageID);
+    if (!hasPermissionToSendWebRequestEvent(tab.get(), request.url(), loadInfo))
+        return;
+
+    auto eventTypes = { WebExtensionEventListenerType::WebRequestOnBeforeRequest, WebExtensionEventListenerType::WebRequestOnBeforeSendHeaders, WebExtensionEventListenerType::WebRequestOnSendHeaders };
+    wakeUpBackgroundContentIfNecessaryToFireEvents(eventTypes, [&] {
+        sendToProcessesForEvents(eventTypes, Messages::WebExtensionContextProxy::ResourceLoadDidSendRequest(tab->identifier(), tab->window()->identifier(), request, loadInfo));
```

Comment:
`tab->window()` can be `null` in rare cases. Will need to check it.

Reply:
I'll adopt something like:

`window ? window->identifier() : WebExtensionWindowConstants::NoneIdentifier;`